packages/editor/src/optimizer/worker.ts

```diff
@@ -0,0 +1,366 @@
+import {
+  PenroseState,
+  compileTrio,
+  finalStage,
+  insertPending,
+  isOptimized,
+  nextStage,
+  resample as penroseResample,
+  start,
+  step,
+} from "@penrose/core";
+import consola from "consola";
+import { Result } from "true-myth";
+import {
+  CompileRequestData,
+  CompileResult,
+  ComputeLayoutRequestData,
+  ComputeLayoutResult,
+  HistoryLoc,
+  MessageRequest,
+  MessageResult,
+  MessageTags,
+  Notification,
+  PollResult,
+  ResampleResult,
+  SizeBoundedMap,
+  StepSequenceID,
+  StepSequenceIDGenerator,
+  StepSequenceInfo,
+  logLevel,
+  notify,
+  respond,
+  stateToLayoutState,
+  taggedErr,
+  taggedOk,
+} from "./common.js";
+
+const log = consola.create({ level: logLevel }).withScope("optimizer:worker");
+const stepSequenceIdGenerator = new StepSequenceIDGenerator();
+const numStepSequencesToKeep = 3;
+
+let historyInfo = new SizeBoundedMap<StepSequenceID, StepSequenceInfo>(
+  numStepSequencesToKeep,
+);
+/**
+ * Varying values sequence for each step sequence. Kept separate from history
+ * info so we don't need to recompute history info on every poll.
+ */
+let historyValues: Map<StepSequenceID, number[][]> = new Map();
+
+/**
+ * Context for optimizer. Varying values only represent the last
+ * values the optimizer saw; refer to `historyValue` to get the appropriate list
+ * for a particular `historyLoc`. This will be non-null always after compilation,
+ * which happens immediately after launch, so it is virtually always safe to use
+ * `penroseState!`.
+ */
+let penroseState: PenroseState | null = null;
+let optimizerShouldStop = false;
+
+/**
+ * Returns a copy of parents varying values (or if null, a copy of the current).
+ * @param stepSequenceInfo
+ */
+const getParentOrCurrentVaryingValues = (
+  stepSequenceInfo: StepSequenceInfo,
+) => {
+  const parentLoc = stepSequenceInfo.parent;
+  return parentLoc === null
+    ? [...penroseState!.varyingValues]
+    : [...historyValues.get(parentLoc.sequenceId)![parentLoc.step]];
+};
+
+/**
+ * Compile, but don't start optimizing until we receive `LabelMeasurementsData`
+ * in a notification. After calling, `penroseState` will be non-null.
+ * @param data
+ */
```

Comment:
Can you or @wodeni remind me why it can't compute the label measurements in the worker itself? For some reason I thought this was possible.

Reply:
mathjax needs a reference to the window : /

Reply:
Sad but makes sense!

Reply:
Yep, copying over an old comment from our thread with @YonahGoldberg summarizing the "polyfill" version of the design:

> when main thread creates worker, send offscreen canvas which worker stores (or if the worker can create it directly, then just do that)
use OffscreenCanvas to allow text measurement to run directly on webworker by making measureTextElement an argument of measureText instead of a global
use liteAdaptor to allow mathjax to run directly on the webworker
also run the renderer on the webworker, generate svg string
send that string across the wire when main thread asks for an update
main thread converts that svg string to dom

There were some weirdness around `OffscreenCanvas` and `LiteAdapter` at the time, but I don't think they can't run on workers in theory.